### PR TITLE
catalog: Avoid clones where possible

### PR DESCRIPTION
**Pull request by jkosh44**
A single update to the catalog is represented using multiple different
Rust structs. Throughout the lifecycle of the update, it needs to be
converted from one type to another. This commit attempts to remove some
of the cloning that is done during these conversions.

Works towards resolving #28722

### Motivation
This PR refactors existing code.

### Tips for reviewer

Review each commit separately. I probably shouldn't have made the first commit, so let me know and I'm happy to revert it.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/catalog/src/durable/persist.rs [hunk 1]
```diff
@@ -694,6 +716,9 @@ impl ApplyUpdate<StateUpdateKindJson> for UnopenedCatalogStateInner {
         current_epoch: &mut FenceableEpoch,
         _metrics: &Arc<Metrics>,
     ) -> Result<Option<StateUpdate<StateUpdateKindJson>>, DurableCatalogError> {
+        // TODO(jkosh44) It's a bit unfortunate that we have to clone all updates to attempt to
+        // convert them into a `StateUpdateKind` and cache a very small subset of them. It would
+        // be better if we could figure out a way not to clone everything.
```
Comment (jkosh44) [inline src/catalog/src/durable/persist.rs:721]: I'm thinking that we can probably inspect the JSON itself to see what the label is and only attempt to convert it based on the label.

**Comment by jkosh44** [comment]
I did some local timings of the various slow catalog related startup sections and got the following:

## `main`

```
catalog init:           1.077803271s,   66.145475ms,    98.593723ms
durable catalog open:   96.450687ms,    91.054152ms,    86.82751ms
open catalog:           9.451893699s,   8.986238331s,   9.074442885s
```

## This PR

```
catalog init:           107.467432ms,   102.740361ms,   64.315825ms
durable catalog open:   285.891382ms,   89.369212ms,    81.850912ms
open catalog:           10.703774926s,  10.493978014s,  9.427781421s
```

It looks like there's no difference, but also these numbers are orders of magnitude smaller than what we see in staging (except for open catalog which is weirdly slower):

```
catalog init:           13.595396853s
durable catalog open:   5.28961681s
open catalog:           2.780077892s
```

so I think the local numbers are probably meaningless.

I want to do some testing in staging, but I also want to get this into the next release. It's possible that this makes no difference, but it would be very surprising if it made things worse.

File: src/catalog/src/durable/objects/state_update.rs [hunk 2]
```diff
@@ -277,19 +278,19 @@ impl StateUpdateKind {
...
+    pub(crate) fn to_serde<D: serde::de::DeserializeOwned>(self) -> D {
         self.try_to_serde().expect("jsonb should roundtrip")
     }
 
     pub(crate) fn try_to_serde<D: serde::de::DeserializeOwned>(
-        &self,
+        self,
     ) -> Result<D, serde_json::error::Error> {
         let serde_value = self.0.as_ref().to_serde_json();
         serde_json::from_value::<D>(serde_value)
```
Comment (jkosh44) [inline src/catalog/src/durable/objects/state_update.rs:296]: It's very possible that these `serde` methods end up cloning the values anyway, it just seemed silly to pass an `&_` when all of the callers had an owned value.

**Review by benesch — APPROVED** [review]
LGTM. I thought about the fencing logic and it made sense, but I'm unfamiliar with this code, so if you have any doubts about that code, it may be worth talking me through it in a bit more detail.

**Comment by jkosh44** [comment]
> LGTM. I thought about the fencing logic and it made sense, but I'm unfamiliar with this code, so if you have any doubts about that code, it may be worth talking me through it in a bit more detail.

I realized I could remove this variant and replace it with an assert, so I just pushed that change which should simplify things a bit.

**Comment by jkosh44** [comment]
Can someone on @MaterializeInc/testing let me know if the Nightly failures are new or known: https://buildkite.com/materialize/nightly/builds/9097. They look scary but unrelated.

**Comment by def-** [comment]
Unrelated it seems.

**Pull request merged**